diff --git a/src/rtl/tc_sram.sv b/src/rtl/tc_sram.sv
index b702a11..476f592 100644
--- a/src/rtl/tc_sram.sv
+++ b/src/rtl/tc_sram.sv
@@ -145,12 +145,14 @@ module tc_sram #(
         for (int unsigned i = 0; i < NumPorts; i++) begin
           if (req_i[i]) begin
             if (we_i[i]) begin
+            `ifndef VERILATOR
               // update value when write is set at clock
               for (int unsigned j = 0; j < BeWidth; j++) begin
                 if (be_i[i][j]) begin
                   sram[addr_i[i]][j*ByteWidth+:ByteWidth] <= wdata_i[i][j*ByteWidth+:ByteWidth];
                 end
               end
+            `endif
             end else begin
               // otherwise update read address for subsequent non request cycles
               r_addr_q[i] <= addr_i[i];
@@ -163,7 +165,9 @@ module tc_sram #(
     // write memory array
     always_ff @(posedge clk_i or negedge rst_ni) begin
       if (!rst_ni) begin
+        `ifndef VERILATOR
         sram <= init_val;
+        `endif
         for (int i = 0; i < NumPorts; i++) begin
           r_addr_q[i] <= {AddrWidth{1'b0}};
           // initialize the read output register for each port
@@ -186,12 +190,14 @@ module tc_sram #(
         for (int unsigned i = 0; i < NumPorts; i++) begin
           if (req_i[i]) begin
             if (we_i[i]) begin
+            `ifndef VERILATOR
               // update value when write is set at clock
               for (int unsigned j = 0; j < BeWidth; j++) begin
                 if (be_i[i][j]) begin
                   sram[addr_i[i]][j*ByteWidth+:ByteWidth] <= wdata_i[i][j*ByteWidth+:ByteWidth];
                 end
               end
+            `endif
             end else begin
               // otherwise update read address for subsequent non request cycles
               r_addr_q[i] <= addr_i[i];
@@ -202,44 +208,116 @@ module tc_sram #(
     end
   end

-// Validate parameters.
-// pragma translate_off
-`ifndef VERILATOR
-`ifndef TARGET_SYNTHESIS
-  initial begin: p_assertions
-    assert ($bits(addr_i)  == NumPorts * AddrWidth) else $fatal(1, "AddrWidth problem on `addr_i`");
-    assert ($bits(wdata_i) == NumPorts * DataWidth) else $fatal(1, "DataWidth problem on `wdata_i`");
-    assert ($bits(be_i)    == NumPorts * BeWidth)   else $fatal(1, "BeWidth   problem on `be_i`"   );
-    assert ($bits(rdata_o) == NumPorts * DataWidth) else $fatal(1, "DataWidth problem on `rdata_o`");
-    assert (NumWords  >= 32'd1) else $fatal(1, "NumWords has to be > 0");
-    assert (DataWidth >= 32'd1) else $fatal(1, "DataWidth has to be > 0");
-    assert (ByteWidth >= 32'd1) else $fatal(1, "ByteWidth has to be > 0");
-    assert (NumPorts  >= 32'd1) else $fatal(1, "The number of ports must be at least 1!");
-  end
-  initial begin: p_sim_hello
-    if (PrintSimCfg) begin
-      $display("#################################################################################");
-      $display("tc_sram functional instantiated with the configuration:"                          );
-      $display("Instance: %m"                                                                     );
-      $display("Number of ports   (dec): %0d", NumPorts                                           );
-      $display("Number of words   (dec): %0d", NumWords                                           );
-      $display("Address width     (dec): %0d", AddrWidth                                          );
-      $display("Data width        (dec): %0d", DataWidth                                          );
-      $display("Byte width        (dec): %0d", ByteWidth                                          );
-      $display("Byte enable width (dec): %0d", BeWidth                                            );
-      $display("Latency Cycles    (dec): %0d", Latency                                            );
-      $display("Simulation init   (str): %0s", SimInit                                            );
-      $display("#################################################################################");
+  // `ifdef VERILATOR
+  for (genvar i = 0; i < NumPorts; i++) begin
+    // update value when write is set at clock
+    for (genvar j = 0; j < DataWidth; j++) begin
+      always_ff @(posedge clk_i or negedge rst_ni) begin
+        if (!rst_ni) begin
+        end else begin
+          if (req_i[i])
+            if (we_i[i])
+              if (be_i[i][j/ByteWidth])
+                sram[addr_i[i]][j] <= wdata_i[i][j];
+        end
+      end
     end
   end
-  for (genvar i = 0; i < NumPorts; i++) begin : gen_assertions
-    assert property ( @(posedge clk_i) disable iff (!rst_ni)
-        (req_i[i] |-> (addr_i[i] < NumWords))) else
-      $warning("Request address %0h not mapped, port %0d, expect random write or read behavior!",
-          addr_i[i], i);
-  end
+  // `endif
+
+// Validate parameters.
+// pragma translate_off
+// `ifndef VERILATOR
+// `ifndef TARGET_SYNTHESIS
+//   initial begin: p_assertions
+//     assert ($bits(addr_i)  == NumPorts * AddrWidth) else $fatal(1, "AddrWidth problem on `addr_i`");
+//     assert ($bits(wdata_i) == NumPorts * DataWidth) else $fatal(1, "DataWidth problem on `wdata_i`");
+//     assert ($bits(be_i)    == NumPorts * BeWidth)   else $fatal(1, "BeWidth   problem on `be_i`"   );
+//     assert ($bits(rdata_o) == NumPorts * DataWidth) else $fatal(1, "DataWidth problem on `rdata_o`");
+//     assert (NumWords  >= 32'd1) else $fatal(1, "NumWords has to be > 0");
+//     assert (DataWidth >= 32'd1) else $fatal(1, "DataWidth has to be > 0");
+//     assert (ByteWidth >= 32'd1) else $fatal(1, "ByteWidth has to be > 0");
+//     assert (NumPorts  >= 32'd1) else $fatal(1, "The number of ports must be at least 1!");
+//   end
+//   initial begin: p_sim_hello
+//     if (PrintSimCfg) begin
+//       $display("#################################################################################");
+//       $display("tc_sram functional instantiated with the configuration:"                          );
+//       $display("Instance: %m"                                                                     );
+//       $display("Number of ports   (dec): %0d", NumPorts                                           );
+//       $display("Number of words   (dec): %0d", NumWords                                           );
+//       $display("Address width     (dec): %0d", AddrWidth                                          );
+//       $display("Data width        (dec): %0d", DataWidth                                          );
+//       $display("Byte width        (dec): %0d", ByteWidth                                          );
+//       $display("Byte enable width (dec): %0d", BeWidth                                            );
+//       $display("Latency Cycles    (dec): %0d", Latency                                            );
+//       $display("Simulation init   (str): %0s", SimInit                                            );
+//       $display("#################################################################################");
+//     end
+//   end
+//   for (genvar i = 0; i < NumPorts; i++) begin : gen_assertions
+//     assert property ( @(posedge clk_i) disable iff (!rst_ni)
+//         (req_i[i] |-> (addr_i[i] < NumWords))) else
+//       $warning("Request address %0h not mapped, port %0d, expect random write or read behavior!",
+//           addr_i[i], i);
+//   end

-`endif
-`endif
+// `endif
+// `endif
 // pragma translate_on
+
+  // Copyright lowRISC contributors.
+  // Licensed under the Apache License, Version 2.0, see LICENSE for details.
+  // SPDX-License-Identifier: Apache-2.0
+
+  /**
+   * Memory loader for simulation
+   *
+   * Include this file in a memory primitive to load a memory array from
+   * simulation.
+   *
+   * Requirements:
+   * - A memory array named `sram`.
+   * - A parameter `DataWidth` giving the memory width (word size) in bit.
+   * - A parameter `NumWords` giving the memory depth in words.
+   */
+
+  `ifndef SYNTHESIS
+  // Task for loading 'sram' with SystemVerilog system task $readmemh()
+  export "DPI-C" task simutil_memload;
+
+  task simutil_memload;
+    input string file;
+    $readmemh(file, sram);
+  endtask
+
+  // Function for setting a specific element in |sram|
+  // Returns 1 (true) for success, 0 (false) for errors.
+  export "DPI-C" function simutil_set_mem;
+  function int simutil_set_mem(input int index, input bit [511:0] val);
+    // Function will only work for memories <= 512 bits
+    if (DataWidth > 512)
+      return 0;
+    if (index >= NumWords)
+      return 0;
+
+    sram[index] = val[DataWidth-1:0];
+    return 1;
+  endfunction
+
+  // Function for getting a specific element in |sram|
+  export "DPI-C" function simutil_get_mem;
+  function int simutil_get_mem(input int index, output bit [511:0] val);
+    // Function will only work for memories <= 512 bits
+    if (DataWidth > 512)
+      return 0;
+    if (index >= NumWords)
+      return 0;
+
+    val                = 0;
+    val[DataWidth-1:0] = sram[index];
+    return 1;
+  endfunction
+  `endif
+
 endmodule
